src/runtime/cpu/cpubackend.cpp

```diff
@@ -54,38 +63,69 @@ dwcpu::CPUBackend::CPUBackend(graph::Graph& graph)
-                             const std::vector<ade::NodeHandle>  & handles) {
+                             const std::vector<ade::NodeHandle>  & handles,
+                                   std::vector<deepworks::Tensor>& mem) {
     for (int i = 0; i < handles.size(); ++i) {
         const auto& d = m_tgraph.metadata(handles[i]).get<graph::Data>();
-        m_mem[d.id] = tensors[i];
+        mem[d.id] = tensors[i];
     }
 }
 
 void dwcpu::CPUBackend::forward(const std::vector<deepworks::Tensor>& inputs,
                                       std::vector<deepworks::Tensor>& outputs) {
     const auto& info = m_tgraph.metadata().get<graph::Info>();
-    bind(inputs , info.in_nhs);
-    bind(outputs, info.out_nhs);
+    bind(inputs , info.in_nhs,  m_mem);
+    bind(outputs, info.out_nhs, m_mem);
 
     auto extract = [this](int id) { return m_mem[id]; };
 
     dwcpu::CPUGraph cpu_graph(m_graph);
     for (auto&& nh : m_ops) {
-        std::vector<deepworks::Tensor> inputs;
-        std::vector<deepworks::Tensor> outputs;
+        std::vector<deepworks::Tensor> ins;
+        std::vector<deepworks::Tensor> outs;
 
-        // NB: Get layer inputs/outputs.
+        // NB: Get layer ins/outs.
         const auto& op = m_tgraph.metadata(nh).get<graph::Op>();
-        std::transform(op.in_ids.begin() , op.in_ids.end() , std::back_inserter(inputs),  extract);
-        std::transform(op.out_ids.begin(), op.out_ids.end(), std::back_inserter(outputs), extract);
+        std::transform(op.in_ids.begin() , op.in_ids.end() , std::back_inserter(ins),  extract);
+        std::transform(op.out_ids.begin(), op.out_ids.end(), std::back_inserter(outs), extract);
 
         // NB: Run layer.
         const auto& cpulayer = cpu_graph.metadata(nh).get<dwcpu::CPUImpl>().impl;
-        cpulayer->forward(inputs, outputs);
+        cpulayer->forward(ins, outs);
     }
 }
 
 void dwcpu::CPUBackend::backward(const std::vector<deepworks::Tensor>& inputs,
-                                       std::vector<deepworks::Tensor>& outputs) {
-    DeepWorks_Assert(false && "Not implemented");
+                                 const std::vector<deepworks::Tensor>& outputs,
+                                 const std::vector<deepworks::Tensor>& grad_outputs) {
+    const auto& info = m_tgraph.metadata().get<graph::Info>();
+    bind(inputs      , info.in_nhs  , m_mem);
+    bind(outputs     , info.out_nhs , m_mem);
+    bind(grad_outputs, info.out_nhs , m_grad);
+
+    auto extract_mem  = [this](int id) { return m_mem[id];  };
+    auto extract_grad = [this](int id) { return m_grad[id]; };
+
+    dwcpu::CPUGraph cpu_graph(m_graph);
+    for (int i = m_ops.size() - 1; i >= 0; --i) {
+        auto nh = m_ops[i];
+
+        std::vector<deepworks::Tensor> inputs;
+        std::vector<deepworks::Tensor> outputs;
+        std::vector<deepworks::Tensor> grad_outputs;
+        std::vector<deepworks::Tensor> grad_inputs;
```

Comment:
Can we reserve capacity?

Reply:
Yes we can, just skip not to increase number of code lines

Reply:
BTW, how many inputs for layer usually used ? I believe `1-6`, so `reserve` capacity is useless